src/coreclr/jit/runtimelookup.cpp

```diff
@@ -0,0 +1,461 @@
+                gtSetStmtInfo(asgFastPathValueStmt);
+                fgSetStmtSeq(asgFastPathValueStmt);
+
+                BasicBlock* sizeCheckBb = nullptr;
+                if (needsSizeCheck)
+                {
+                    // Dynamic expansion case (sizeCheckBb is added and some preds are changed):
+                    //
+                    // prevBb(BBJ_NONE):                    [weight: 1.0]
+                    //
+                    // sizeCheckBb(BBJ_COND):               [weight: 1.0]
+                    //     if (sizeValue <= offsetValue)
+                    //         goto fallbackBb;
+                    //     ...
+                    //
+                    // nullcheckBb(BBJ_COND):               [weight: 0.8]
+                    //     if (*fastPathValue == null)
+                    //         goto fallbackBb;
+                    //
+                    // fastPathBb(BBJ_ALWAYS):              [weight: 0.64]
+                    //     rtLookupLcl = *fastPathValue;
+                    //     goto block;
+                    //
+                    // fallbackBb(BBJ_NONE):                [weight: 0.36]
+                    //     rtLookupLcl = HelperCall();
+                    //
+                    // block(...):                          [weight: 1.0]
+                    //     use(rtLookupLcl);
+                    //
+
+                    sizeCheckBb = fgNewBBbefore(BBJ_COND, nullcheckBb, true);
+                    sizeCheckBb->bbFlags |= BBF_INTERNAL;
+
+                    // sizeValue = dictionary[pRuntimeLookup->sizeOffset]
+                    GenTreeIntCon* sizeOffset = gtNewIconNode(runtimeLookup.sizeOffset, TYP_I_IMPL);
+                    assert(lastIndOfTree != nullptr);
+                    GenTree* sizeValueOffset = gtNewOperNode(GT_ADD, TYP_I_IMPL, lastIndOfTree, sizeOffset);
+                    GenTree* sizeValue       = gtNewOperNode(GT_IND, TYP_I_IMPL, sizeValueOffset);
+                    sizeValue->gtFlags |= GTF_IND_NONFAULTING;
+
+                    // sizeCheck fails if sizeValue <= pRuntimeLookup->offsets[i]
+                    GenTree* offsetValue =
+                        gtNewIconNode(runtimeLookup.offsets[runtimeLookup.indirections - 1], TYP_I_IMPL);
+                    GenTree* sizeCheck = gtNewOperNode(GT_LE, TYP_INT, sizeValue, offsetValue);
+                    sizeCheck->gtFlags |= GTF_RELOP_JMP_USED;
+                    gtSetEvalOrder(sizeCheck);
+                    Statement* sizeCheckStmt = fgNewStmtFromTree(gtNewOperNode(GT_JTRUE, TYP_VOID, sizeCheck));
+                    sizeCheckStmt->SetDebugInfo(debugInfo);
+                    gtSetStmtInfo(sizeCheckStmt);
+                    fgSetStmtSeq(sizeCheckStmt);
+                    fgInsertStmtAtEnd(sizeCheckBb, sizeCheckStmt);
+                }
+
+                // Replace call with rtLookupLclNum local
+                call->ReplaceWith(gtNewLclvNode(rtLookupLclNum, call->TypeGet()), this);
```

Comment:
This should rather use `BashToLclVar`, but if you address the feedback above by reusing this call directly then you'll probably need `gtGetParent` instead (with some extra logic for when the call is the root of the statement).